resonance_tester: Resonance testing and input shaper auto-calibration

dmbutyugin opened the pull request:
Signed-off-by: Dmitry Butyugin <dmbutyugin@google.com>

dmbutyugin commented:
@KevinOConnor To simplify the review and integration process, I decided to split the feature into several pull requests:

- shaper autocalibration module and stand-alone scripts
- `resonance_tester` module that can run the vibration test and that uses `shaper_calibrate` module
- documentation

I hope that's fine with you.

KevinOConnor commented:
Thanks.  Splitting up the commits is fine, but I'd prefer to see all the commits in order to review.  (It's hard to review code without also being able to see the ultimate code using that code.)

Separately, even if we add analysis to klippy, I'd still prefer to not add graphing code to klippy.  User interface stuff like graphs tend to have a lot churn and I think the focus for klippy should be real-time motion, timing, kinematics, etc.

Cheers,
-Kevin

dmbutyugin commented:
> Thanks. Splitting up the commits is fine, but I'd prefer to see all the commits in order to review. (It's hard to review code without also being able to see the ultimate code using that code.)

OK, I've put all the relevant code in this pull request, and updated its name.


> Separately, even if we add analysis to klippy, I'd still prefer to not add graphing code to klippy. User interface stuff like graphs tend to have a lot churn and I think the focus for klippy should be real-time motion, timing, kinematics, etc.

I've made the changes and moved all the plotting-related code to `scripts/`.

dmbutyugin commented:
BTW, I've made the wiring diagrams myself and put them into this pull request, instead of directing users to the external repo.

KevinOConnor commented:
Thanks!  I suspect it's going to take me a couple of days to review this though.  I am looking forward to merging it.

Cheers,
-Kevin

dmbutyugin commented:
Thanks! Sure, I understand, it's a lot of new code and docs.

Sineos commented:
I gave this PR a test. I did run:
1. `TEST_RESONANCES AXIS=X CSV_NAME=resonances.csv RAW_NAME=raw_data.csv `
2. `TEST_RESONANCES AXIS=Y CSV_NAME=resonances.csv RAW_NAME=raw_data.csv `
3. `SHAPER_CALIBRATE`

After the `TEST_RESONANCES` there was a "Wait for calculations.." output 2 times. It though seemed that nothing happened.

**Results:**
![resonances_x](https://user-images.githubusercontent.com/4459312/95254400-23f27500-0820-11eb-809b-b7b32242f1f7.png)

![resonances_y](https://user-images.githubusercontent.com/4459312/95254447-340a5480-0820-11eb-96fa-4caf0e171116.png)


`SHAPER_CALIBRATE` recommended:
- shaper_type_x = mzv, shaper_freq_x = 25.0 Hz
- shaper_type_y = 3hump_ei, shaper_freq_y = 69.2 Hz

While I would agree with the Y-axis recommendation (bit aggressive IMO but highly efficient in the reduction), I wonder about the X-axis recommendation:  Wouldn't a shaper around 90Hz do a better job?

Finally, how do I graph the `SHAPER_CALIBRATE` results as seen in the documentation?

After all: Great job. Highly appreciated

**Attached:**
klippy.log and raw data.
[test_20201006.zip](https://github.com/KevinOConnor/klipper/files/5336481/test_20201006.zip)

dmbutyugin commented:
@Sineos, thanks!

> I gave this PR a test. I did run:
> 
>     1. `TEST_RESONANCES AXIS=X CSV_NAME=resonances.csv RAW_NAME=raw_data.csv `
> 
>     2. `TEST_RESONANCES AXIS=Y CSV_NAME=resonances.csv RAW_NAME=raw_data.csv `
> 
>     3. `SHAPER_CALIBRATE`
> 
> 
> After the `TEST_RESONANCES` there was a "Wait for calculations.." output 2 times. It though seemed that nothing happened.

It should have written `resonances*.csv` files, which are the results of those "Wait for calculations..".

> `SHAPER_CALIBRATE` recommended:
> 
>     * shaper_type_x = mzv, shaper_freq_x = 25.0 Hz
> 
>     * shaper_type_y = 3hump_ei, shaper_freq_y = 69.2 Hz
> 
> 
> While I would agree with the Y-axis recommendation (bit aggressive IMO but highly efficient in the reduction), I wonder about the X-axis recommendation: Wouldn't a shaper around 90Hz do a better job?

TBH, this is a case near the threshold where the automatic calibration may switch from MZV to 2HUMP_EI (in SHAPER_CALIBRATE case the thresholds were in favour of MZV, but in case of raw data from TEST_RESONANCES it would choose 2HUMP_EI at 37.5 Hz). But it's really about the values being near the switching threshold:
| ![raw_data_x](https://user-images.githubusercontent.com/53491797/95258371-f0b2e480-0825-11eb-9142-7b82c195dcdd.png) | ![calibration_data_x](https://user-images.githubusercontent.com/53491797/95258392-f8728900-0825-11eb-961a-35e8d984010e.png) |
|:--:|:--:|

Note that the graph_accelerometer.py just plots the frequency response as-is, but the shaper calibration additionally normalizes the frequency response by frequency (so the amplitude of peaks differs for shaper calibration). Thus making the resonances at 90-100 Hz less prominent (and it is generally true that they require less compensation).

But, this is an interesting case in other regard, so thanks for sharing it! Right now the code assumes that using MZV is generally better than EI, 2HUMP_EI, etc., because it usually gives less smoothing. And to choose more 'aggressive' input shaper that input shaper must provide significant reduction in vibrations. However, in this example it is less obvious (even though MZV@25 Hz still gives less smoothing than 2HUMP_EI@37.5 Hz). So maybe some improvements are due. But I suspect the improvements may need to be not in shaper selection itself, but elsewhere.

BTW, which `max_accel` value have you been using for this test? 7000 as suggested in the docs?

> Finally, how do I graph the `SHAPER_CALIBRATE` results as seen in the documentation?

You cannot do this from `SHAPER_CALIBRATE` command anymore. I removed this possibility following Kevin's request (to not plot the charts from within Klipper directly). It does save CSV files by default (calibration_data*.csv) though. You can still do it from the command line (either on RPi or on host), running something like (with appropriate paths):
```
$ klipper/scripts/calibrate_shaper.py raw_data_x_20201006_203207.csv -o calibration.png
$ klipper/scripts/calibrate_shaper.py calibration_data_x_20201006_205037.csv -o calibration.png
```
If you run this on host, you can omit `-o calibration.png` - it can plot the chart in a window. And you may need to force some version of Python; I, for instance, run it as `$ python3 klipper/scripts/calibrate_shaper.py ...` (so the scripts are Python3-compatible).

Sineos commented:
> It should have written `resonances*.csv` files, which are the results of those "Wait for calculations..".

It did. Maybe a message like "Writing values to /tmp" or an indication when the processing has finished would be more transparent.

> BTW, which `max_accel` value have you been using for this test? 7000 as suggested in the docs?

Yes, 7000 for both `max_accel` and `max_accel_to_decel`.

> You can still do it from the command line (either on RPi or on host), running something like (with appropriate paths)

Works nicely, thanks. I think that this is a nice functionality. Although it is mentioned in the docs, it could be a tad clearer.

KevinOConnor reviewed (commented):
Thanks!  At a high-level it looks good to me.  See my comments below - mostly minor things.  I do think we should address the toolhead.max_accel_to_decel and filename stuff prior to merging.

-Kevin

Arksine commented on klippy/toolhead.py:
```diff
@@ -505,7 +505,7 @@ def get_status(self, eventtime):
                      'max_velocity': self.max_velocity,
                      'max_accel': self.max_accel,
-                     'max_accel_to_decel': self.requested_accel_to_decel,
+                     'max_accel_to_decel': self.max_accel_to_decel,
```
Thanks for the heads up.  Mainsail uses this data inform the UI of the existing velocity/accel settings.  I don't think changing it to the internal value would break things if you choose to change it.   Worst case scenario we can report both values as you suggest.  I will run it by Meteyou.

dmbutyugin replied:
In principle, the name of the exported variable does not change, but only the value does. As I mentioned in the comment below, this change actually 'fixes' the export to the actual value used by Klipper (i.e. if the user requested too high value and that value was 'trimmed' to the lower value). But I do not think this particular change is very important, I made it more for consistency, of reporting and to avoid the confusion, so I can revert it if you feel strongly against it. @Arksine in light of this, what do you think Mainsail should report?

KevinOConnor replied:
FWIW, I tend to think of "requested max accel to decel" to be the "real" value, and the internal "trimmed max accel to decel" to be an internal implementation detail.  That is, if the requested accel to decel is greater than the max accel, then it doesn't actually impact the print, but it's still a "valid value".

Looking closer, the new code doesn't actually call the `get_status()` method at all.  So, I think it would be better to separate this out into a separate PR (if needed at all).

dmbutyugin replied:
I moved these changes to a separate PR #3422, it is probably best to continue the discussion there. But TBH I do not have a very strong opinion about 'max accel to decel', I just made this change for consistency with the rest of reporting, and because the user may be more interested in the actually used value (even if it is an implementation detail, if a user requested to change max accel to decel, they probably want to know whether this request succeeded or not).

dmbutyugin commented on docs/G-Codes.md:
```diff
@@ -642,14 +642,50 @@ section is enabled:
+is enabled:
+- `MEASURE_AXES_NOISE`: Measures and outputs the noise for all axes of all
+  enabled accelerometer chips.
+- `TEST_RESONANCES AXIS=<axis> [CSV_NAME=<csv_file>] [RAW_NAME=<raw_cvs_file>]
+  [FREQ_START=<min_freq>] [FREQ_END=<max_freq>] [HZ_PER_SEC=<hz_per_sec>]`:
+  Runs the resonance test in all configured probe points for the requested
+  axis (X or Y) and measures the acceleration using the accelerometer chips
+  configured for the respective axis. If `RAW_NAME` is set, then the raw
+  accelerometer data is written into a file or a series of files
+  `/tmp/<raw_csv_name>_<axis>_<point>_YYYYMMDD_HHMMSS<raw_csv_ext>`.
+  If `CSV_NAME` is set, it is treated as a template
+  `<csv_file> = <csv_name>+<csv_ext>`; the frequency response is calculated
+  and written into `/tmp/<csv_name>_<axis>_YYYYMMDD_HHMMSS<csv_ext>` file.
+  At least one of the `CSV_NAME` or `RAW_NAME` parameters must be provided.
```
I was thinking that it might be too easy to overwrite an existing file if the user forgets to update the file name (in repeated measurements). So, for example, `AXIS=Y RAW_NAME=resonances.csv` will produce a file `/tmp/resonances_x_YYYYMMDD_HHMMSS.csv` or, if the multiple test points were configured, e.g. to `/tmp/resonances_x_100.000_100.000_20.000_YYYYMMDD_HHMMSS.csv`. At least, some template expansion is likely necessary (as opposed to just take `RAW_NAME` value) in case multiple test points were configured (an alternative would be to concatenate the measurements into a single file, but that's not great, in my opinion). So, Kevin, what would you prefer: make this command resemble `ACCELEROMETER_MEASURE` (but then what to do with raw data from multiple test points?), maybe try to clarify the description and/or add a some examples how the file name will be expanded?

Separately, I am going to add a confirmation message following @Sineos suggestion: just print the file name of the file written. So maybe that will also help to avoid the confusion.

KevinOConnor replied:
I guess I'd say use a hard-coded template expansion, but allow the user only specify a "name" that (if not specified) defaults to YYYYMMDD_HHMMSS.  So, for example, if the user specified `NAME=foo` then an output file might look like: `/tmp/resonances_x_100.000_100.000_20.000_foo.csv`.

dmbutyugin replied:
I made the changes along those lines.

dmbutyugin commented on docs/Measuring_Resonances.md:
````diff
@@ -0,0 +1,339 @@
+
+However, if you connected two accelerometers simultaneously and configured them
+in the following manner:
+```
+[adxl345 adxl345_x]
+cs_pin: ...
+
+[adxl345 adxl345_y]
+cs_pin: ...
````
That's good to know! Maybe then I can mention exactly that: 2 chips can be used, as long as they are wired to different boards (or different SPI interfaces)? Or do you think it might be better to skip this feature for now?

KevinOConnor replied:
FWIW, I think the feature is fine.  Just wanted to let you know about that limitation of the adxl345 chips.

dmbutyugin replied:
I explicitly wrote about this limitation in the docs now.

dmbutyugin commented on klippy/extras/resonance_tester.py:
```diff
@@ -0,0 +1,283 @@
+                              "%.6f (x), %.6f (y), %.6f (z)" % (
+                                  axis, vx, vy, vz))
+
+    def get_filename(self, name_tmpl, axis=None, point=None):
+        # Cleanup name_tmpl and get only the name
+        name_tmpl = os.path.split(os.path.normpath(name_tmpl))[1]
+        base, ext = os.path.splitext(name_tmpl)
+        time_suffix = time.strftime("_%Y%m%d_%H%M%S")
+        if axis:
+            base += '_' + axis
+        if point:
+            base += "_%.3f_%.3f" % (point[0], point[1])
+        base += time_suffix
+        return os.path.join("/tmp", base + ext)
```
> I'm leery of allowing users to specify filenames from the g-code command line - both because I'm leery of an accident overriding an important system file and because I'm leery of a malicious gcode file being used to corrupt a system.

Note that the code currently runs `os.path.split(os.path.normpath(name_tmpl))[1]` to at least attempt to prevent a user from doing something weird and writing files outside of `/tmp/` folder. And in terms of malicious gcode: I honestly think that the rouge gcode can do much more damage than just overwriting an important file and corrupting a system: it can cause a physical damage to the printer and even maybe set it on fire. So in general, it is a very bad idea to take untrusted gcode and run it on your printer. Though I agree that sometimes it is necessary to do that for debugging.   

> As such, I'd prefer to enforce a file name template determined by the source code - for example `/tmp/calibration_data_<name>.csv` where the user may only specify `<name>` and the code verifies that `<name>` does not contain any weird characters. For example, see how the adxl345.py code performs this check.

In general, forcing a fixed template with some short <name> override might be a good idea though, because it may also simplify the documentation how it is supposed to work.

KevinOConnor replied:
> Note that the code currently runs os.path.split(os.path.normpath(name_tmpl))[1]

Ah, I saw that, but wasn't sure what it did.  FWIW, I'd also prevent "weird" characters in the filename like '$', '`', and similar.

I agree that running malicious g-code would be very dangerous.  However, I would still feel better if we didn't add "system level privilege escalation" to the list of dangers.  I guess I don't see a real-world use case for "custom filenames" and I'd just feel better if we didn't "open that can of worms".

dmbutyugin replied:
> However, I would still feel better if we didn't add "system level privilege escalation" to the list of dangers. I guess I don't see a real-world use case for "custom filenames" and I'd just feel better if we didn't "open that can of worms".

OK, makes sense, I made changes to the output naming scheme similar to ACCELEROMETER_MEASURE so this should not longer be an issue.

dmbutyugin commented on klippy/toolhead.py:
```diff
@@ -572,8 +572,8 @@ def cmd_SET_VELOCITY_LIMIT(self, gcmd):
                "square_corner_velocity: %.6f"% (
-                   max_velocity, max_accel, self.requested_accel_to_decel,
-                   square_corner_velocity))
+                   self.max_velocity, self.max_accel, self.max_accel_to_decel,
+                   self.square_corner_velocity))
```
I can make this a separate PR. The reason I added it to this PR initially was because:

1. Currently Klipper reports back the parameters from the request, even if it didn't apply them. Then the message is not particularly helpful. I have seen quite some confusion from the users (myself included at first), who believed that their limit was applied. This can be especially problematic if the user wanted to, but forgot to change the limit in their config (or did that but forgot to run `RESTART`) - they will believe they are doing the right thing without a good way to detect their mistake. So I think it would be great to fix that.

2. Resonance tester code issues a command to increase the maximum acceleration (and set accel-to-decel = max_accel); here arguably the user didn't request any limit change directly, so it is best to report the actually configured new limits. I was thinking that this can lead to some extra confusion if the user configured, say, `max_accel = 5000` because they know this is how much their printer can handle, and then they'll see that max_accel was set to 9000 (which would be 75 * 120), and can freak out with 'My printer will fall apart from this!' thought. That's why I put this change into this PR (but I'm happy to commit it separately).

Now why I made the change to `get_status` method: that method is used in logging, AFAICT. If it isn't updated, but this code is changed, this will lead to inconsistent messages in the logs. Please let me know how you think about updating this code and leaving `get_status` method be.

KevinOConnor replied:
I understand.  I think it would be best to do these changes separately from resonance tester - as resonance testing is new functionality and these changes alter existing functionality.  Since they're not directly tied to each other, I think it would be easier to commit one and then the other.  (And if there is a regression it will then be easier to rollback one without having to rollback both.)

dmbutyugin replied:
I moved these changes to a separate PR #3422.

dmbutyugin commented:
@KevinOConnor Thanks for the review! I already started addressing your comments (didn't push the changes to the review branch yet though), but I wanted to clarify with you a couple of things, so I added replies to some of your comments already.

dmbutyugin commented on docs/Measuring_Resonances.md:
```diff
@@ -0,0 +1,339 @@
+Measuring Resonances
+====================
+
+This branch provides a support for ADXL345 accelerometer, which can be used to
```
Fixed.

dmbutyugin commented on docs/Measuring_Resonances.md:
```diff
@@ -0,0 +1,339 @@
+measure resonance frequencies of the printer for different axes, and auto-tune
+[input shapers](Resonance_Compensation.md) to compensate for resonances.
+Note that using ADXL345 requires some soldering and crimping. Also note that
+only Raspberry Pi setups have been tested at this time.
```
Reworded.

dmbutyugin commented on config/example-extras.cfg:
```diff
@@ -582,6 +587,44 @@
+#   Determines the speed of the test. When testing all frequencies in range
+#   [min_freq, max_freq], each second the frequency increases by hz_per_sec.
+#   Small values make the test slow, and the large values will decrease the
+#   precision of the test.
```
Added the default values explicitly (with units).

dmbutyugin commented on docs/Measuring_Resonances.md:
```diff
@@ -0,0 +1,339 @@
+
+Generated CSV files show power spectral density of the vibrations depending on the
+frequency. Usually, the charts generated from these CSV files are pretty much
+self-explanatory, with the peaks corresponding to the resonance frequencies:
```
Removed that part about being self-explanatory :)

dmbutyugin commented on klippy/extras/adxl345.py:
```diff
@@ -190,33 +222,43 @@ def finish_measurements(self):
+        self.start_measurements()
+        reactor = self.printer.get_reactor()
+        eventtime = reactor.monotonic()
+        while not self.raw_samples:
+            eventtime = reactor.pause(eventtime + .1)
```
Makes sense. Added a timeout.

dmbutyugin commented:
@KevinOConnor I think I made the suggested changes.

Separately, I want to make some further small changes to better account the shaper smoothing in the shaper autocalibration, but I think it would be best to make them in a separate PR (since this will be an improvement of an existing functionality).

KevinOConnor commented on docs/G-Codes.md:
```diff
@@ -642,14 +642,54 @@ section is enabled:
+- `MEASURE_AXES_NOISE`: Measures and outputs the noise for all axes of all
+  enabled accelerometer chips.
+- `TEST_RESONANCES AXIS=<axis> OUTPUT=<resonances,raw_data>
+  [CSV_NAME=<csv_name>] [RAW_NAME=<raw_name>] [FREQ_START=<min_freq>]
```
Very minor, but wouldn't it be simpler if there was just a single `NAME` parameter?  I can't see a case where a user would wish to use a "name" differently between the two.

dmbutyugin replied:
That's a good idea. Done.

KevinOConnor commented on scripts/graph_accelerometer.py:
```diff
@@ -42,6 +39,126 @@ def plot_accel(data, logname):
+    ax.plot(freqs, py, label='Y', alpha=0.6)
+    ax.plot(freqs, pz, label='Z', alpha=0.6)
+
+    ax.xaxis.set_minor_locator(matplotlib.ticker.AutoMinorLocator())
+    ax.yaxis.set_minor_locator(matplotlib.ticker.AutoMinorLocator())
+    ax.xaxis.set_minor_locator(matplotlib.ticker.AutoMinorLocator())
+    ax.yaxis.set_minor_locator(matplotlib.ticker.AutoMinorLocator())
```
Is the double setting (here and on line 121-124) actually correct?

dmbutyugin replied:
Yep, it was an accumulated copy-paste error, apparently.

KevinOConnor commented on scripts/graph_accelerometer.py:
```diff
@@ -52,20 +169,42 @@ def setup_matplotlib(output_to_file):
-    data = parse_log(args[0])
+    datas = [parse_log(fn) for fn in args]
+
+    if options.output and os.path.splitext(options.output)[1].lower() == '.csv':
+        write_frequency_response(datas, options.output)
+        return
```
I think it would be better to add a separate option (eg, `--csv myfile.csv`) than to overload the meaning of the `-o` option.

dmbutyugin replied:
Well, right now the user may specify different extensions already: `-o resonances_x.png` or `-o resonances_x.svg` or `-o resonances_x.pdf`, etc., and the output will be different according to the extension. Of course, you cannot really 'write' a picture into CSV file, but I thought that, from the user perspective, it would feel natural to similarly change the extension and get essentially the same thing in a different format. However, to look less hacky and more natural, I changed the code to also support spectrogram, and added explicit error messages if `--raw` or `--compare` was requested.

KevinOConnor reviewed (commented):
Thanks!  Looks good to me.  I have a few minor comments below.

Do you wish to merge this before the release?  (I'm okay with that if you want to.)

-Kevin

dmbutyugin commented:
> Thanks! Looks good to me. I have a few minor comments below.
> 
> Do you wish to merge this before the release? (I'm okay with that if you want to.)
> 
> -Kevin

I think it would make sense to include it into the upcoming Klipper release - after all, a big chunk is already integrated (e.g. accelerometer support). Then the feature will be complete if this was integrated.

KevinOConnor commented on scripts/graph_accelerometer.py:
```diff
@@ -56,7 +57,7 @@ def calc_specgram(data, axis):
     window = np.kaiser(M, 6.)
     def _specgram(x):
-        return matplotlib.mlab.specgram(
+        return mlab.specgram(
```
FWIW, I'd avoid this change as matplotlib has "funky" order of import rules wrt calling `matplotlib.use('Agg')`.  It's easier to use the longer mlab name explicitly.

-Kevin

KevinOConnor commented:
Thanks.  Looks good to me.  I'll squash and merge tomorrow.  (If you'd like me to preserve the commit history, then that's fine too - just let me know.)

-Kevin

dmbutyugin commented:
> Thanks. Looks good to me. I'll squash and merge tomorrow. (If you'd like me to preserve the commit history, then that's fine too - just let me know.)
> 
> -Kevin

Thanks! No, I do not think the commit history is of any particular interest, let's commit it as a single commit.

KevinOConnor commented:
Thanks!

-Kevin
